src/coreclr/src/vm/processdiagnosticsprotocolhelper.cpp

```diff
@@ -114,6 +136,134 @@ bool ProcessInfoPayload::Flatten(BYTE * &lpBuffer, uint16_t &cbSize)
     return fSuccess;
 }
 
+void EnvironmentHelper::PopulateEnvironment()
+{
+    CONTRACTL
+    {
+        NOTHROW;
+        GC_TRIGGERS;
+        MODE_PREEMPTIVE;
+    }
+    CONTRACTL_END;
+
+    if (Environment != nullptr)
+        return;
+
+    // env block is an array of strings of the form "key=value" delimited by null and terminated by null
+    // e.g., "key=value\0key=value\0\0";
+    LPWSTR envBlock = GetEnvironmentStringsW();
+    LPWSTR envCursor = envBlock;
+    // first calculate the buffer size
+    uint32_t nWchars = 0;
+    uint32_t nEntries = 0;
+    while(*envCursor != 0) 
+    {
+        uint32_t len = static_cast<uint32_t>(wcslen(envCursor) + 1);
+        nEntries++;
+        nWchars += len;
+        envCursor += len;
+    }
+
+    // serialized size value can't be larger than uint32_t
+    S_UINT32 size(sizeof(uint32_t) + (nEntries * sizeof(uint32_t)) + (nWchars * sizeof(WCHAR)));
+    // if the envblock size value is larger than a uint32_t then we should bail
+    if (!size.IsOverflow())
+    {
+        // copy out the Environment block
+        LPWSTR tmpEnv = new (nothrow) WCHAR[nWchars + 1];
+        if (tmpEnv != nullptr)
+        {
+            memcpy(tmpEnv, envBlock, (nWchars + 1) * sizeof(WCHAR));
+            Environment = tmpEnv;
+        }
+        _nEnvEntries = nEntries;
+        _nWchars = nWchars;
+    }
+
+    FreeEnvironmentStringsW(envBlock);
+
+    return;
+}
+
+uint32_t EnvironmentHelper::GetEnvironmentBlockSize()
+{
+    LIMITED_METHOD_CONTRACT;
+
+    S_UINT32 size(sizeof(uint32_t) + (_nEnvEntries * sizeof(uint32_t)) + (_nWchars * sizeof(WCHAR)));
+    return size.IsOverflow() ? sizeof(uint32_t) : size.Value();
+}
+
+bool EnvironmentHelper::WriteToStream(IpcStream *pStream)
+{
+    CONTRACTL
+    {
+        NOTHROW;
+        GC_TRIGGERS;
+        MODE_PREEMPTIVE;
+        PRECONDITION(pStream != nullptr);
+        PRECONDITION(Environment != nullptr);
+    }
+    CONTRACTL_END;
+
+    // Array<Array<WCHAR>>
```

Comment:
In the overflow case this still writes lots of data even though GetEnvironmentBlockSize() specified only 4 bytes would be written. Separately I would assume in the overflow case we'd want to return an error in the header and write 0 bytes.

Reply:
The logic should leave `_nEnvEntries` set to `0`, so the payload would end up saying the continuation has 4 bytes and those 4 bytes should be the number 0 since there won't be any entries in the array.  I could modify it so that the continuation length in the message payload is 0 instead and let the user treat that as an error.  Would that be preferable?